source/game/kart/KartMove.cc

```diff
@@ -1384,6 +1402,166 @@ void KartMove::landTrick() {
     activateBoost(KartBoost::Type::TrickAndZipper, duration);
 }
 
+/// @addr{0x8058498C}
+void KartMove::enterCannon() {
+    init(true, false);
+    physics()->clearDecayingRot();
+    m_boost.resetActive();
+    state()->setBoost(false);
+
+    m_jumpPadMinSpeed = 0;
+    m_rampBoost = 0;
+    state()->setRampBoost(false);
+    m_ssmtCharge = 0;
+    m_ssmtLeewayTimer = 0;
+    m_ssmtDisableAccelTimer = 0;
+    m_offroadInvincibility = 0;
+
+    state()->setBoostOffroadInvincibility(false);
+
+    dynamics()->reset();
+
+    state()->setHop(false);
+    state()->setInCannon(true);
+    state()->setSkipWheelCalc(true);
+    state()->setCannonStart(false);
+
+    const auto [cannonPos, cannonDir] = cannonPosRot();
+    const EGG::Vector3f kartPos = this->pos();
+    m_cannonEntryPos = kartPos;
+    m_cannonEntryOfs = cannonPos - kartPos;
+    this->m_cannonSmth = m_cannonEntryOfs.normalise();
+    m_cannonEntryOfs.normalise();
+    m_dir = m_cannonEntryOfs;
+    m_vel1Dir = m_cannonEntryOfs;
+    m_cannonOrthog = EGG::Vector3f::ey.perpInPlane(m_cannonEntryOfs, true);
+    m_cannonProgress.setZero();
+}
+
+/// @addr{0x80584D58}
+void KartMove::calcCannon() {
+    // WARNING: all these variable names are BAD and INACCURATE because I DONT KNOW WHAT IM DOING
+    // DO NOT be misled by the INACCURATE AND MISLEADING VARIABLE NAMES
+    // this includes ALL cannon-related variables, such as m_cannonEntryPos and m_cannonSmth
+
+    const auto [cannonPos, cannonDir] = cannonPosRot();
+    EGG::Vector3f unk0 = cannonPos - m_cannonEntryPos - m_cannonProgress;
+    EGG::Vector3f unk1 = unk0;
+    const f32 unk1Mag = unk1.normalise();
+    unk0.y = 0;
+    unk0.normalise();
+    EGG::Vector3f local94 = m_cannonEntryOfs;
+    local94.y = 0;
+    local94.normalise();
+    m_speedRatioCapped = 1.0f;
+    // this is actually a 3x3 matrix
+    EGG::Matrix34f local58;
+    local58.makeOrthonormalBasis(unk1, EGG::Vector3f::ey);
+    EGG::Vector3f up = local58.multVector33(EGG::Vector3f::ey);
+    m_smoothedUp = up;
+    m_up = up;
+
+    // exiting cannon
+    if ((unk1Mag < 30.0f || local94.dot(unk0) <= 0.0f) && state()->isInCannon()) {
+        state()->setInCannon(false);
+        state()->setSkipWheelCalc(false);
+        dynamics()->setIntVel(m_speed * m_cannonEntryOfs);
+        return;
+    }
+
+    local58.mulRow33(0, EGG::Vector3f::ez);
+    local58.mulRow33(1, EGG::Vector3f::ez);
+    m_speed = m_baseSpeed;
+    const System::MapdataCannonPoint *cannonPoint =
```

Comment:
Convention thus far has been to use `auto` when it simplifies out a different namespace specifier
```suggestion
    const auto *cannonPoint = System::CourseMap::Instance()->getCannonPoint(state()->cannonPointId());
```
Likewise I would probably suggest doing this for `cannonParams`